multi-machine jobs and locking docs

aaannz opened the pull request.

coolo commented on docs/WritingTests.asciidoc:
```diff
@@ -219,3 +219,151 @@ $ /usr/share/openqa/script/clone_job.pl --from https://openqa.opensuse.org  --ho
+[id="mm-tests"]
+
+Scenarios requiring more than one system under test (SUT), like High Availability testing, are covered as multi-machine tests (MM tests) in section.
+
```
in this section? "covered in section" == CSI ;)

coolo commented on docs/WritingTests.asciidoc:
```diff
@@ -219,3 +219,151 @@ $ /usr/share/openqa/script/clone_job.pl --from https://openqa.opensuse.org  --ho
+Scenarios requiring more than one system under test (SUT), like High Availability testing, are covered as multi-machine tests (MM tests) in section.
+
+OpenQA approaches this problem by assigning dependencies between individual jobs. This means following:
+
```
if following is followed by a : it's a nop

sysrich replied:
'This means the following:' would be valid.

coolo commented on docs/WritingTests.asciidoc:
```diff
@@ -219,3 +219,151 @@ $ /usr/share/openqa/script/clone_job.pl --from https://openqa.opensuse.org  --ho
+
+* _everything needed for MM tests must be running as a test job_ (or you are on your own), even support infrastructure (custom DHCP, NFS,
+etc. if required), which in principle is not part of the actual testing, must have a defined test suite so a test job can be created
+* OpenQA scheduler makes sure _tests are started as a group_ and in right order, _canceled as a group_ if some dependencies are violated and _cloned as
```
cancelled

coolo replied:
hmm. while we use cancelled everywhere in the code, it seems to be british english?

sysrich replied:
Well I think we should have British English as the 'standard English' of the openSUSE Project, so I agree with the code where we're using cancelled :)

aaannz replied:
Agree. Which *spell does KDE use? :D

coolo commented on docs/WritingTests.asciidoc:
```diff
@@ -219,3 +219,151 @@ $ /usr/share/openqa/script/clone_job.pl --from https://openqa.opensuse.org  --ho
+
+1. documenting the dependencies and order between individual tests
+2. synchronization between individual tests
+3. actual technical realization (i.e. https://github.com/os-autoinst/os-autoinst/blob/master/doc/networking.md[custom networking])
```
hmm, shouldn't we move this docu into here?

sysrich replied:
If the custom networking documentation is only relevant to multi-machine jobs, then yes, I think we should
(and I cant think of a time you'd need to play with the networking besides getting more than one VM to talk to another, which is, by definition, a multi-machine job)

aaannz replied:
I think it should be here, as in openqa docs, but not necessarily tided to MM. You still can do sort-of-MM without openQA MM support (as we did with autoyast last year - custom separate support server). I will do that and convert it to asciidoc.

coolo commented on docs/WritingTests.asciidoc:
```diff
@@ -219,3 +219,151 @@ $ /usr/share/openqa/script/clone_job.pl --from https://openqa.opensuse.org  --ho
+* OpenQA provides _locking server for basic synchronization_ of tests (e.g. wait until DHCP/PXE server is ready), but the _correct usage of locks is 
+test designer job_ (beware deadlocks).
+
+In short, writing multi-machine tests add few more layers of complexity:
```
I think it's "adds" but I'm unsure

sysrich replied:
Correct '...tests adds a few...' would be most correct

coolo commented on docs/WritingTests.asciidoc:
```diff
@@ -219,3 +219,151 @@ $ /usr/share/openqa/script/clone_job.pl --from https://openqa.opensuse.org  --ho
+
+There are 2 types of dependencies: +CHAINED+ and +PARALLEL+:
+
+* +CHAINED+ describes when one test case depends on another and are run sequentially, i.e. KDE test suite is run after and only after Installation test suite
```
"are run" -> "both are run"

sysrich replied:
actually.. "are run sequentially" is 'valid' as it implies the 'both' (you cant run a single thing sequentially)
"both are run sequentially" is certainly a little better though

coolo commented on docs/WritingTests.asciidoc:
```diff
@@ -219,3 +219,151 @@ $ /usr/share/openqa/script/clone_job.pl --from https://openqa.opensuse.org  --ho
+* +CHAINED+ describes when one test case depends on another and are run sequentially, i.e. KDE test suite is run after and only after Installation test suite
+is successfully finished and canceled if fail.
+
+To define +CHAINED+ dependency add variable +START_AFTER_TEST+ with the name(s) of test suite(s) after which selected test suite is supposed to run.
```
I'm missing the information how to specify the information how to specify the plural. comma seperated list, right?

coolo replied:
and we're missing the fact that this only affects jobs on the same machine

sysrich replied:
As this only affects jobs on the same machine, what does it have to do with Multi-Machine jobs?

From a user perspective, should this stuff about Chained dependencies make more sense to be in part of the 'regular' Docs, with just the stuff about PARALLEL being appropriate for MM jobs?

coolo commented on docs/WritingTests.asciidoc:
```diff
@@ -219,3 +219,151 @@ $ /usr/share/openqa/script/clone_job.pl --from https://openqa.opensuse.org  --ho
+
+* +PARALLEL+ describes MM test, test suites are run at the same time and managed as a group.
+
+To define +PARALLEL+ dependency, use +PARALLEL_WITH+ variable with the name(s) of test suite(s) which acts as a parent suite to selected test suite.
```
do I have to give all others in every machine or will dhcp PARALLEL_WITH dhcpd be enough?

coolo replied:
and what is the meaning of a parent in parallel?

sysrich replied:
Can I nest parents? eg. Can I have a setup like MM_suite_1 with setting PARALLEL_WITH='ha' followed by MM_suite_2 with setting PARALLEL_WITH='MM_suite_1'

I have no idea why I'd want to do this, it just got me thinking..

aaannz replied:
@coolo Once is enough, the wording of variable is a little misleading I think. What you actually say by PARALLEL_WITH=dhcpd is 'I need test suite dhcpd to be running during my run'. But @nadvornik will know more I guess.

@sysrich AFAIK Yes, you can nest them but then expect interesting behavior when you abort&clone the parent (service provider, the one others depend on... something like that). Basically you'll need 3D model to visualize what Vladimir's algorithm will do with that.

coolo commented on docs/WritingTests.asciidoc:
```diff
@@ -219,3 +219,151 @@ $ /usr/share/openqa/script/clone_job.pl --from https://openqa.opensuse.org  --ho
+Worker requirements
+^^^^^^^^^^^^^^^^^^^
+
+Beware OpenQA worker requirements. +CHAINED+ dependency requires only one worker, since dependent jobs will run only after the first one finish.
```
"beware of", but I think the whole statement is just fillup. If you want to add a warning, mark it explicitly at the end of the section so it stands out for a casual reader. People reading every sentense *will* be aware of it.

coolo commented on docs/WritingTests.asciidoc:
```diff
@@ -219,3 +219,151 @@ $ /usr/share/openqa/script/clone_job.pl --from https://openqa.opensuse.org  --ho
+On the other hand +PARALLEL+ dependency requires at least 2 workers for simple scenarios.
+
+Keep in mind parent job _must be running until all
+children finish_, else scheduler will cancel child jobs once parent is done.
```
Again: what is a parent and what are children if they are parallel?

sysrich replied:
I think you need to better explain how PARENT and CHILD test suites need to be written
The requirement for PARENT test_suites to always take longer than its children is certainly important.
Are there any other requirements, guidelines, steps that a PARENT or CHILD test suite should contain?

coolo commented on docs/WritingTests.asciidoc:
```diff
@@ -219,3 +219,151 @@ $ /usr/share/openqa/script/clone_job.pl --from https://openqa.opensuse.org  --ho
+.+CHAINED+ - i.e. test basic functionality before going advanced - requires 1 worker
+----
+A <- B <- C
+
```
I would expect the arrows the other way around if it's first A then B - just saying :)

coolo commented on docs/WritingTests.asciidoc:
```diff
@@ -219,3 +219,151 @@ $ /usr/share/openqa/script/clone_job.pl --from https://openqa.opensuse.org  --ho
+
+Define test suite A
+and then define B,C,D with PARALLEL_WITH=A
+----
```
So A is the parent because it's not PARALLEL, but still all 4 are scheduled at once? I think 'parent' is a very bad word for this.

But in general your examples clear up some points of above, pointing to them from above might help.

coolo commented on docs/WritingTests.asciidoc:
```diff
@@ -219,3 +219,151 @@ $ /usr/share/openqa/script/clone_job.pl --from https://openqa.opensuse.org  --ho
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+OpenQA provides locking server through lock API. To use lock API import +lockapi+ package (_use lockapi;_) in your test file.
+Lock API provides three functions: +mutex_create+, +mutex_lock+, +mutex_unlock+. Each of these functions take one parameter, name of the lock.
```
I think a : before 'name of lock' is easier to read

coolo commented on docs/WritingTests.asciidoc:
```diff
@@ -219,3 +219,151 @@ $ /usr/share/openqa/script/clone_job.pl --from https://openqa.opensuse.org  --ho
+
+By default, when mutex lock does not exist, lock is considered locked and +mutex_lock+ blocks. After lock is created by +mutex_create+, lock
+is automatically unlocked. +mutex_unlock+ unlocks the lock.
+
```
I fail to see the the need for mutex_create from this explanation. Why not mutex_unlock creating the lock automatically and mutex_lock waiting for the lock to be created or unblocked?

aaannz replied:
There was an idea to use mutex_unlock as a form of blocking is_locked mechanism. So that it blocks when locked, but after unlocked it does not change the lock. Unfortunately as I go through the code this does not work, yet. Will fix.
The following example illustrates exactly that ... not. :(

coolo commented:
Feel free to ignore everything you don't agree with

aaannz commented:
I addressed almost all comments, except moving the networking part to openqa docs. I'll do that in separate PR.
And I left the arrows in the original direction as they symbolize dependencies.
I also "redraw" parallel graphs, added HA example, added parent/children meaning, reworded locks api and that's all I think

coolo commented:
coverage remained the same - not counting documentation coverage obviously :)
